compose/cli/log_printer.py

```diff
@@ -41,7 +42,7 @@ def no_color(text):
         for color_func, container in zip(color_funcs, self.containers):
             generator_func = get_log_generator(container)
             prefix = color_func(build_log_prefix(container, prefix_width))
-            yield generator_func(container, prefix, color_func)
+            yield generator_func(container, prefix, color_func, self.cascade_stop)
```

Comment:
So if we want to support the case 3 you mention in the issue, I think we'd need to make a couple more changes.

Instead of raising an exception here, I think we would just let this exit. Then from the call in `compose/cli/main.py` we can check the flag, and if it's set, we can run the shutdown from there.